New guide for installing new Home Assistent on Synology

Mr-Meorzipan opened the pull request:
This guide will help users in detail how to install any new Home Assistant version in the foreseeable future (until the next bump in the road that may cause incompatibility).

Synology only provide Python 3.5.1, which is not compatible with Home Assistant 0.65.0 or later (basically  a complete standstill), so I have created this documentation based on mine and others research from https://community.home-assistant.io/t/python-3-5-3-on-synology/46372

The key is installing and compiling newer Python, some fixes for SSL and broken components (which were previously broken too).
The rest is a more optimized installation that will may cause less issues to arise.

**Description:**


**Pull request in [home-assistant](https://github.com/home-assistant/home-assistant) (if applicable):** home-assistant/home-assistant#<home-assistant PR number goes here>

## Checklist:

- [ ] Branch: `next` is for changes and new documentation that will go public with the next [home-assistant](https://github.com/home-assistant/home-assistant) release. Fixes, changes and adjustments for the current release should be created against `current`.
- [ ] The documentation follows the [standards][standards].

[standards]: https://developers.home-assistant.io/docs/documentation_standards.html

Mr-Meorzipan committed:
New instuctions for installing new Home Assistent on Synology

This documentation will guide users in detail how to install any new Home Assistant version in the foreseeable future.

Synology only provide Python 3.5.1, which is not compatible with Home Assistant 0.65.0 or later (basically  a complete standstill), so I have created this documentation based on mine and others research from https://community.home-assistant.io/t/python-3-5-3-on-synology/46372/27

The key is installing a compiling and installing newer Python and some fixes for SSL and broken components (which were previously broken too).
The rest is a more optimized installation that will cause less issues to arise (aside from a static virtual python environment for HASS).
The Python package installation will also be left untainted thanks to that, which is nice when upgrading Synology, since often it will erase any tainted modification on packages (eg, erasing Home Assistent).

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
-<p class='note warning'>
-Synology only provide Python 3.5.1, which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. You can manually specify the version of Home Assistant to install, for example to install version 0.64.3 you would do `./python3 -m pip install homeassistant==0.64.3`
+<p class="note warning">
+Synology only provide <a href="https://www.synology.com/nl-nl/dsm/packages/py3k">Python 3.5.1</a>, which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. You can manually specify the version of Home Assistant to install, for example to install version 0.64.3 y ...
```
Please use Markdown to format links
```raw
[Python 3.5.1](https://www.synology.com/nl-nl/dsm/packages/py3k)
```

Mr-Meorzipan replied:
Does Markdown work inside html tags, like Paragraph?
When I try to use Markdown formatted hyperlinks, these are not parsed in the preview.
Instead to compensate, I used hyperlink tags.

In the case Markdown is not supported in this way and html hyperlinks is not accepted, should I remove hyperlinks altogether and only add the URL in "Thread for more info: Python >=3.5.3 on Synology."?
I will not be able to link to the components pages, so should I remove the info about which components for example will work? That info is useful so the user has a idea what this guide fixes, compared to the previous guide.

I could also instead remove the Paragraph tag and just use normal text, but the user might think with only the warning note, that is applicable to the new guide on first look (which with a note to the update info it catches the eye, so the user would read that first instead of skimming).

fabaff replied:
It should work.

Mr-Meorzipan replied:
Alright, I will be changing it all to Markdown then.

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
+</p>
+
+<p class="note">
+Update: You can install a more recent home assistant (eg, 0.84.2) by manually compiling a python3 package using the <a href="https://github.com/SynoCommunity/spksrc">spksrc compilation framework</a>. The new installation guide will be a little different though, but the upside is most failing components will now work, such as “<a href="/components/discovery/">Discovery</a>”, “<a href="/components/ ...
```
It's `Home Assistant` and `Python 3`. And please use `e.g.,` as we agreed to use it that way across the documentation.

Mr-Meorzipan replied:
The "python3" name refers to the package name for Synology, not the project+version.
https://www.synology.com/nl-nl/dsm/packages/py3k
The package name internally is "py3k" though, so I in that sense the package name is not really correct anyway, so I will change it based on your direction.
The rest I will apply too.

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
+
+Using the [Synology webadmin](https://www.synology.com/en-global/knowledgebase/DSM):
+* Install compiled Python3 package using the [Synology Package Center](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)
+* Create “homeassistant” user and add to the “users” group
```
Please use `"` for quotes.

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
 
-SSH onto your synology & login as admin or root
+In this guide we are going to use a Linux based Operating System such as [Ubuntu](https://www.ubuntu.com/), [Mint](https://linuxmint.com/) or [Solus](https://getsol.us) to minimize difficulty. You will also need [Docker](https://docs.docker.com/install/).
+You can think of [Docker](https://docs.docker.com/install/) as a way more powerful [Chroot](https://help.ubuntu.com/community/BasicChroot) and containers as ready made mini operating systems.
```
I think that this is nice to know but doesn't add value to the guide.

Mr-Meorzipan replied:
I am assuming Windows users and people on different Linux distributions might follow the guide, so it is nice to let them know what system they could run to make everything work nicely.
I could remove the Docker sentence though.

Mr-Meorzipan replied:
The steps displaying how to install Docker already has a header for the system it runs on.
Currently the header for Ubuntu, Mint and derivatives has a hyperlink to the Docker page and the Solus URL to the Solus homepage.
For Solus there is no specific Docker page.
Should I change the Ubuntu hyperlink to the appropriate homepage hyperlink, or leave as is?

Mr-Meorzipan replied:
I have changed as you requested and removed that part, should I change the header to instead include the homepages, leave as Docker installation page hyperlink or no hyperlink at all?

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
 
- - Log in with your own administrator account
- - Switch to root using:
+---
```
Remove this.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+```bash
+# sudo apt-get install docker-ce
+```
+---
````
Not needed.

Mr-Meorzipan replied:
This is a integral part of the guide, as a docker container is needed.
The alternative is the user installing build-essential and all sorts of dev packages, preferable in a virtual machine, at which point this is going to have too many steps and going to be user unfriendly, which is why I prefer Docker.

This is all needed to compile new Python 3 Synology package, which I already made steps to make it copy and paste friendly.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+# sudo apt-get install docker-ce
+```
+---
+### {% linkable_title Installing Docker on [Solus](https://getsol.us) %}
````
Remove the Solus section and all parts that refer to Solus. If a user is using Solus it's usually an experienced user who an adapt the Ubuntu guide.

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
 
-Create hass-daemon file using the following code (edit the variables in uppercase if necessary)
 
+## {% linkable_title Using the Synology webadmin: %}
```
```suggestion
## {% linkable_title Using the Synology webadmin %}
```

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+```
+Create a virtual Python environment where we will install Home Assistant into.
+This will leave the Python package untouched, makes Home Assistant work better and installable/updateable without sudo.
+Note: After you made a virtual Python envoriment, you cant relocate the folder, else it will break. 
````
```suggestion
Note: After you made a virtual Python environment, you can't relocate the folder, else it will break. 
```

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
+## {% linkable_title Installing Home Assistant %}
+
+After the Python 3 package has been installed, open terminal and open SSH to the synology.
+Replace “<i>user</i>” with your Synology user and “x.x.x.x” with the its IP-Adress.
```
```suggestion
Replace “<i>user</i>” with your Synology user and “x.x.x.x” with the its IP address.
```

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -175,68 +383,25 @@ case $1 in
 ```
 
-Create links to python folders to make things easier in the future:
+## {% linkable_title Controlling Home Assistent %}
````
```suggestion
## {% linkable_title Controlling Home Assistant %}
```

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+# deactivate
+```
+
+Create a file named “hass-daemon” in the “homeassistant” Shared-Folder with the script below as it's content.
````
```suggestion
Create a file named “hass-daemon” in the “homeassistant” Shared-Folder with the script below as its content.
```

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+Note: After you made a virtual Python envoriment, you cant relocate the folder, else it will break. 
+Make a new one if you need to change it's name, the path or Shared-Folder.
+```bash
+# /volume1/@appstore/python3/bin/python3 -m venv /volume1/homeassistant/venv-hass
````
`#` should only be used if the command is executed by root.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+After the Python 3 package has been installed, open terminal and open SSH to the synology.
+Replace “<i>user</i>” with your Synology user and “x.x.x.x” with the its IP-Adress.
+```bash
+# ssh user@x.x.x.x
````
`#` should only be used if the command is executed by root.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+Refresh package lists:
 ```bash
-$ sudo -i
+# sudo apt-get update
````
`#` should only be used if the command is executed by root. Use `$` otherwise.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+Run the container, this will make your terminal run inside the Docker container:
 ```bash
-# ./python3 -m pip install homeassistant==0.64.3
+# docker run -it -v ~/spksrc:/spksrc synocommunity/spksrc /bin/bash
````
Can a regular user run Docker?

Mr-Meorzipan replied:
By regular user, do you mean a less technical user or a less privileged user?
If by less privileged user, the use of superuser is needed when installing packages and using Systemd to activate docker service, the rest has been written so it can be run as normal user.

The "docker run" command is done without superuser by following the guide and running:
[
To enable your user account to manage Docker without superuser privileges, add your user account to the “docker” group:
```bash
# sudo gpasswd -a YOUR_USER_ACCOUNT docker
```
]

If you mean less technical user, the guide has been made to be very simple and explaining, as you could for the most part just copy paste commands and everything works, until the parts you need to edit files or add slightly modify commands to insert user or IP.
The use of Home Assistant already sets a point were you need some technical knowledge (in the sense of terminal, "scary" commands and the sort).

fabaff replied:
I was talking about a less privileged system user.

Mr-Meorzipan replied:
If it is better, I can remove the step for adding the user to the 'docker' group (which allowed that regular user to run docker) and instead require superuser for 'docker run'.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+# make setup
+# cd spk/python3
 ```
+<p class='note'>
````
I don't think that this needs to be a note.

fabaff commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,287 @@
+It should be named something like “python3_armada370-6.1_3.5.5-7.spk”, of course with a possibly different arch and version.
 
+Now you need to extract your “cross compiled module package” files which you added earlier.
+For ease of use this guide will use p7zip for extracting from these archives.
```
Why not use `tar/zip` which is likely to be in the default package set?

Mr-Meorzipan replied:
For simplicity, so the user can simply copy paste a command and stuff just works.
I know no default program that can rename files inside archives without extracting.
7-Zip can and does it fast and quietly and since the guide describes installing Docker and Git, installing p7zip shouldn't be a problem.

The user also does not have to use Git and 7-Zip if he does not want/can, since the user can use the "Clone/Download" button on Github and rename all files as described inside the ".whl" files manually using any zip archiving program.

If I should change this anyway, I could write a command which unpacks, renames and repacks the file.
I could also add a special dd command that renames all files without extracting...nah better to not.

fabaff replied:
I was just asking. It make sense to use the the tools which provide the easiest way to do something.

Mr-Meorzipan replied:
If it makes that part better, I have made most of the commands to use usually found tools in the system.
The part for renaming the archives I have still made use of p7zip, but made it a command which will rename all files required to be renamed in all '.whl' archives found in it's current directory instead of presuming a certain file.

Mr-Meorzipan replied:
Hello, I made a command which makes use of the usually included commands in a linux based system to do the renaming job, however I do feel like it is more efficient to just use the p7zip based command, as it's shorter, simply renames files inside archives without unpacking and is quite a bit faster.
Though that would usually require installing p7zip.

This one will do the same kind of big operation, but instead does the following: Do a operation on all .whl archives, unzip archive, rename files, create zip, delete tempdir, repeat.
It runs in the current directory and is basically a copy paste should just work kind of command.
```sh
rand=$RANDOM; for module in *.whl; do unzip "$module" -d "temp$rand" && find "temp$rand" -name "*x86_64-linux-gnu*" -type f | while read -r file; do mv "$file" "$(echo $file | sed "s/x86_64-linux-gnu/arm-linux-gnueabihf/")"; done && rm "$module" && (cd "temp$rand" && zip -r0 "../$module" ./) && rm -r "temp$rand"; done
```
Should I replace the 7z based command with this?

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+# 7z l "Module-Packages/pycryptodome-3.7.2-cp35-none-any.whl" | grep "x86_64-linux-gnu" | cut -c54- | while read -r file; do 7z rn "Module-Packages/pycryptodome-3.7.2-cp35-none-any.whl" "$file" "$(echo $file | sed "s/x86_64-linux-gnu/arm-linux-gnueabihf/")"; done
+```
+<p class='note'>
+The 7z <b>rn</b> (eg, rename) parameter was included from 7-Zip 9.30, in the case you are pitifully stuck on a older version, do as follows.
````
```suggestion
The 7z <b>rn</b> (eg, rename) parameter was included from 7-Zip 9.30, in the case you have older version, do as follows.
```

Mr-Meorzipan replied:
I can change the sentence to:
[
in the case your distribution only has a older version available, do as follows.
]
I this acceptable?

fabaff replied:
Fine with me.

fabaff commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,287 @@
+```
+<p class='note'>
+The 7z <b>rn</b> (eg, rename) parameter was included from 7-Zip 9.30, in the case you are pitifully stuck on a older version, do as follows.
+Extract the .whl (these are zip archives), rename all files as described above, then rearchive as zip (ofcourse the same full name with .whl and without .zip).
````
```suggestion
Extract the .whl (these are zip archives), rename all files as described above, then rearchive as zip (use the same full name with `.whl` as extension and not `.zip`).
```

Mr-Meorzipan replied:
Is there a reason you used backticks instead of quotes?

arsaboo replied:
Backticks allows the code to be formatted properly.

Mr-Meorzipan replied:
In what cases should is using backticks appropriate?
it is not clear to me yet the use case between quotes and backticks.

Right now I understand backticks here are to show part of the filename, but should I be using backticks in all places were something might be used in commands or filenames?

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

frenck commented:
Just wondering... Why not guide users to install Home Assistant via Docker on Synology?

CharlieDowns commented:
Docker is not supported on certain Synology's because of lack of things like lack 64bit support, not enough Memory (e.g. 500MB), lack of CPU power.
For example, Docker is not supported on DS115J and DS216, so you are required to run it directly on the filesystem, running the local Python 3 installation.
For supported models, look at: https://www.synology.com/nl-nl/dsm/packages/Docker

But as you know, Python 3 was not updated by Synology in a long time, which is bad for Home Assistant since new versions require new Python 3 versions.
So this guide helps users in making new Python 3 packages, installing and configuring the Synology correctly and installing Home Assistant.

Essentially if everything works well, you could basically just copy paste commands and everything just works (excluding some commands which needs different user or content).

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

frenck commented:
Thanks for clearing that up @CharlieDowns, nevertheless, just because a device "could" run Home Assistant, that doesn't mean we should provide/advertise a guide for it on our end.

e.g., You could root an android phone and run Home Assistant on it... 

Personally, I'm kinda in split if this should be added or not (in general).

CharlieDowns commented:
Hello frenck,
When this has been setup on the Synology, it does run very well, I am even using on mine (DS115J) every day with ESP8266 Milight and Zigbee2MQTT with Xiaomi Aqara devices and have noticed no bugs or slowdowns.

You are however right though in the sense it is complicated to setup.
It is douable though, I tested it multiple times on different Synology's, the complicated part in the setup comes from the Python compilation or the instruction when a Python module fails when configuring components.

There is really no other way to run Home Assistant on Synology though except for running a older version, or you would need run it in Docker, which the cheapest compatible costs around €300.
The real problem is really the Synology development team not updating their stuff (Synology, if your reading this, please update your py3k package, we are already waiting years).

Maybe the guide could be redone in some parts?

For example, the part of installing Docker, I am not sure if it is really needed to install from their repository, I could add only instruction to install from Ubuntu's repository, making it only 2 commands and remove the gpasswd command.
I could also replace the whole 7z part were some files in archives need to be renamed with a command that uses usually included commands, but 7zip is the only program I know that can rename without unpacking.
The part were you make a virtual Python environment I could also replace with installing Home Assistant into the Python 3 installation directory.

frenck commented:
I understand, the question in this case: Should we support it? Because adding this to our main site/documentation, might imply that.

A better place for these kinds of things (IMHO) is a blog somewhere, tbh for 35 bucks, you'll get a much better option (Raspberry Pi) (again IMHO).

I leave this one up to @fabaff, he reviewed it as well. I was just noticing this one and dropped in my 2 cents.

CharlieDowns commented:
I get what you mean and adding this might mean allowing more of the same.
Personally I can't really say to whether include it or not, but this does indeed feel a more advanced and involved then the usual guides.

I can only make it shorter and simpler and tell people to run a command, copy some text in a file somewhere and it just works.
I could even add a warning at the top that running on a Raspberry Pi will give a better experience, but if people really want to run it, to know what there getting into.

I also feel even though the part of the component, cross compiled python module packages is always fixable, might still cause more forum questions to arise since it is the most confusing part.

Update: I remember SynoCommunity used to host a pretty old version of Python 3, I checked again and seems they have updated it to version 3.5.6.
If I include this in the guide, meaning adding the [SynoCommunity repository](https://synocommunity.com/#easy-install), Clicking the [option to allow trusted publishers](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/configure), looking in the "Community" tab and installing Python 3 the guide becomes a whole lot easier and allows updates.
The spksrc compilation framework also gets the sources from SynoCommunity, so no differences in version.
The package looks like it comes with most needed modules, except it misses "cryptography", but it seems the SynoCommunity maintainers have a [thread](https://github.com/SynoCommunity/spksrc/issues/3115) asking which modules to add to their requirements_all.txt for Home Assistant.

They also host a package for Home Assistant version 82.1, but does not seem be kept updated very often. Currently it is not available for me on DSM 6.2.


Anyway, this all does feel more DIY then anything else. It works really well when setup, but the general newbie might get stuck. I should probably add a note letting users know what they are getting into.
To get most of all components working, it seems cross compilation is still needed in case these are not included in the Python3 SynoCommunity package.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

CharlieDowns commented:
Hello, I edited the guide so it is shorter and has less steps.
it still has the compilation part, but may feel better now.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

ymartin59 commented:
Hello. I have updated spksrc repository based on Allan Clark contribution https://github.com/SynoCommunity/spksrc/pull/3462
I cannot prepare HA binary package with all wheels that may be required by any enabled HA components, but I have planned to add them "on-demand".

CharlieDowns commented:
Hello ymartin59, 
That is understandable, if it was me, I would include the modules for components which are most likely to be used and point users to the github page or some form for inclusion of components.

Then every time when you make a new Home Assistant package, you could use a sample configuration.yaml which has all the components you want to include. You would then start the Home Assistant binary and see which modules fail to install, these changed required modules you add/change on requirements.txt, cross compile and include in the package.
Should be fairly straight forward and I would see only 1-2 hours needed for that.

If your interested, there seems also to be a Python module specifically for porting modules to embedded systems without having to compile Python itself called "crossenv".

Once the package is ready, I can include it in the guide since it is way simpler, but I will also keep the current method, as when people want to keep their Home Assistant current or add their own components, they can do it them self without needing to wait for inclusion.

ymartin59 commented:
@CharlieDowns Thanks for hint. I will look at crossenv but I fear to spend hours to compile native libraries required for each python module in HA component - as I did for homekit to get gmpy compiled against cross-compiled gmp, mpfr and mpc... My next step is openzwave with udev rule to gain access to Zwave USB-dongle.
Have you got a proposal of configuration file as a significant sets of most "used" components by HA users?

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

CharlieDowns commented:
@ymartin59, Honestly, I don't. I do not spend that much time in the forums really, so I do not know which components are most likely to be used by people.

I can imagine adding modules which fail to install for any components that you see in the "Featured" tab of the [components](https://www.home-assistant.io/components/) page of Home Assistant page is a good start.
You could also look around to see what seems recently to be popular or open a thread on Home Assistant forum and ask there, as you then ask the people who might use your package, as they most likely end up on the forum.

I do not know how you test which modules to add to your requirements.txt, but my method is as follows:
 * Install component and see if it fails
 * Check log for "Unable to install package *Module_Name*"
 * Install that module with `pip` and see which required modules failed
 * Install these required modules and see which fails with "Failed building wheel for *module_name*"
 * Add these package to requirements.txt

Usually I only get 1 or 2 packages from that, for example Homekit needs "HAP_python", but it fails because "curve25519-donna" and "ed25519" need to be cross compiled.
For this procedure I need only 1 one compile and takes me around 30 minutes on a SSD [MX300] and 4rd gen I5 CPU.

When I decide to add a few components to Home Assistant, I do the same thing once.

The procedure I take is written on the guide I wrote, you could check it out when you have time.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

chickenandpork commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,196 @@
-<p class='note warning'>
-Synology only provide Python 3.5.1, which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. You can manually specify the version of Home Assistant to install, for example to install version 0.64.3 you would do `./python3 -m pip install homeassistant==0.64.3`
+<p class="note warning">
+Synology only provides [Python 3.5.1](https://www.synology.com/en-global/dsm/packages/py3k), which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. If you want to run newer Home Assistant versions, you will need to make a new Python 3 package, which will be de ...
```
...Until Synology offers ...

"...need to make a new Python 3 package" needs to be "...install the new Python 3 package from SynoCommunity"

Maybe remove the space between "python" and "3"; additionally, "/var/packages/python3/target/bin/pip3 install homeassistant==0.64.3" works.

Using "./python3" (ie "current directory" requires previous context to not change; I would suggest full pathnames so that the reader doesn't need to track down what directory he or she needs to be in.

Mr-Meorzipan replied:
Thanks @chickenandpork, I did not pay much attention to that part since it was from the original guide, even though I should have.
I do wonder if I should include a part of the old guide at the end of the document under some header like "Previous guide" or something like that, then link to that from the warning note.

Mr-Meorzipan replied:
Btw, is it better to link from "/volume1/@appstore/python3" or "/var/packages/python3" and if so why?
Currently all python3 binaries in the guide are linked from "@appstore", but I do not mind changing that to try make things more future-proof.
Might it be possible users have a different directory such as "/volume2" for "@appstore" or is this a almost non existent case?

chickenandpork replied:
First, it should be `/volume1/@appstore/python3`

I would suggest `/var/packages/python3` rather than `/volume1/@appstore/python3` because `volume1` is chosen by the user (I have two volumes, for example) so if the user chooses to install in `volume2`, the resulting path is a softlink from `/var/packages/python3/target` to `/volume2/@appstore/python3/` -- notice how `/var/packages/python3` remains correct, but `/volume1/` is now incorrect?  `/var/` links seem more predictable in this really rare case.

Mr-Meorzipan replied:
For me the location does not really matter, as long as it is predictable.
Is there a big chance Synology might change "/var/packages/python3" to some different location in the future? If not, I would be using "/var" in a flash.

I mean, if a new DSM causes a change of the var path, I would prefer "/volume1", since it is easy to tell a user to change a number because of different volume.
Telling to find somewhere in "/var" is harder and confusing.

chickenandpork commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,196 @@
+```
+Fork and clone spksrc, this will make a directory named "spksrc".
+```bash
+$ git clone https://github.com/SynoCommunity/spksrc.git
````
If the command on line 72 assumes you're cloning into `~/spksrc` then perhaps you should:

`git clone https://github.com/SynoCommunity/spksrc.git ~/spksrc`

Mr-Meorzipan replied:
Hmm, in case the terminal is in a different directory, the cloned repository would be in a different directory.
Thanks, will change.

chickenandpork commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,196 @@
+cryptography==2.3.1
 
-Use PIP to install Homeassistant package 0.64.3
+## Cross compilation requirement for installing "Warrent" module (Needed by "Cloud" Component)
```
s/Warrent/Warrant/ ?

Mr-Meorzipan replied:
Because warrant and warrent sound the same to me, these kind of mistakes happen so often...
Will correct.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

chickenandpork commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,196 @@
+$ tar -x -f $pyspk -C /tmp package.tgz; gzip -df /tmp/package.tgz
+$ tar -x -f /tmp/package.tar --wildcards share/wheelhouse/$file --strip=2
+```
+Inside some of the `.whl` files (These are zip archives) you need to rename all files containing the text "**x86_64-linux-gnu**" to "**arm-linux-gnueabihf**", this is only required for ARM based Synology's.  
````
nit: "Synology's" is possessive as though a Synology owns something; "Synologies" would be plural except that "Synology" is a proper noun, hence "Synologys" (odd) or "Synology DSMs" (fairly neutral) is most applicable.

Mr-Meorzipan replied:
Synology DSMs would refer to the operating system, which in this case seems to be the most fitting.
English is not my native language, but looking at some noun rules, should "DSMs" not be "DSMs'"?

chickenandpork replied:
You're correct, it would be "Synology DiskStations" or "Synology Products".  This is like the "Lego-vs-Legos" argument, where "Legos" is not actually plural of "Lego": it would be "Lego Bricks", "Lego toys", or just "Lego" (pluralizing like "sheep", "moose", "fish")

Synology public relations refers to their products as "Synology Products" or "Synology NAS" (I guess using "NAS" as plural for "NAS" :) )

Mr-Meorzipan replied:
NAS is more user friendly, DSMs' sounds weird, I will be taking NAS.

Mr-Meorzipan replied:
I decided to remove the "#Extracting cross compiled packages" part since after a little testing I saw that installing the `.whl` manually is not needed, which means the edits discussed in this conversation are invalidated.
Must have left it around because in the beginning I installed Home Assistant in virtual python environment.

If you want to read the lastest bit of that part, you can find it in [this commit](https://github.com/home-assistant/home-assistant.io/pull/7861/commits/07efb626821f34c2439331537fad040a40e8bb40).

Oh well...

chickenandpork commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,196 @@
-# chown homeassistant /volume1/homeassistant 
-# chmod 755 /volume1/homeassistant
-# cd /volume1/homeassistant
+# sudo docker run -it -v ~/spksrc:/spksrc synocommunity/spksrc /bin/bash
```
`sudo docker run -it -v ~/spksrc:/spksrc synocommunity/spksrc  -w /spksrc /bin/bash`

You can almost one-shot this.  Two command lines would be:

```
sudo docker run -it --rm -v ~/spksrc:/spksrc -w /spksrc synocommunity/spksrc  make setup
sudo docker run -it --rm -v ~/spksrc:/spksrc -w /spksrc synocommunity/spksrc  make -C spk/python3 arch-XXXX-6.1
```

Mr-Meorzipan replied:
Why do you specify the DSM version in the second command?
Wouldn't it be better leave it to the default setting when executing `make setup`?

Good idea, better than a potential area of confusion when users forget their terminal is active in the docker container.

Could probably shorten by multiple commands with bash -c parameter.
```bash
sudo docker run -it --rm -v ~/spksrc:/spksrc -w /spksrc synocommunity/spksrc bash -c 'make setup && make -C spk/python3 arch-XXXX'
```

Mr-Meorzipan replied:
Although, might adding the version be better in the case a user runs a older version of DSM (e.g., 6.1)? I mean in the case his NAS is no longer supported and no new DSM is available?
Might new DSM become incompatible because of setting a DSM version?

Mr-Meorzipan replied:
When the user executes `docker run`, his container is removed because of `--rm`.
In case the compilation is cancelled or fails, his docker container would also be removed?
Is it not better to let the user decide to remove the containers?

I might even have to add  a container remove command anyway in case the user runs `docker pull`, but the download fails, filling "/var/lib/docker/overlay" with corrupted containers.

I tried looking for a command to delete corrupted containers, but `docker rm` did not seem to erase these. I was able to erase these by stopping docker service and deleting "/var/lib/docker", but what are the odds the user is actually using Docker for other purposes and then has to re-download all his Docker containers...?

Mr-Meorzipan replied:
Never mind, tested again, seems the thing with the Docker containers were just a placebo effect and it cleans-up just fine.

Though I still oppose to removing the docker container when running the compilation command, a better solution is adding a separate command for removing the docker container, that way the user won't have to re-download the container on every run and it is easier to decide if he wants to remove the container and the user won't be confused if he runs the command a second time and gets a error because the docker container was removed.

chickenandpork commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,196 @@
+#### {% linkable_title Compiling the Python 3 package %}
+Make the Python 3 package for your Synology model, please replace "arch-**XXXX**" by the apporiate architecture of your Synology. For a list of architectures, look at this [list of architectures](https://github.com/SynoCommunity/spksrc/wiki/Architecture-per-Synology-model) accepted by spksrc. Depending on your computer, compilation may take a hour or more (Significantly less if you have a SSD and ...
 ```bash
-# mkdir /volume1/homeassistant
````
`mkdir -m 0755 /volume1/homeassistant`

Mr-Meorzipan replied:
Making a directory manually was a part of the old guide.
The instruction for the current guide is to make a "homeassistant" Shared-Folder.

chickenandpork commented:
...and that happens even if you've seen the text a dozen times; a new set
of eyes can find little nit-picky issues.  The really nuanced ones are
really up to you to accept or refuse -- it's your PR in the end

On Sun, Jan 6, 2019 at 1:59 AM CharlieDowns <notifications@github.com>
wrote:

> *@CharlieDowns* commented on this pull request.
> ------------------------------
>
> In source/_docs/installation/synology.markdown
> <https://github.com/home-assistant/home-assistant.io/pull/7861#discussion_r245502714>
> :
>
> >
> -Use PIP to install Homeassistant package 0.64.3
> +## Cross compilation requirement for installing "Warrent" module (Needed by "Cloud" Component)
>
> Because warrant and warrent sound the same to me, these kind of mistakes
> happen so often...
> Will correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant.io/pull/7861#discussion_r245502714>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACNonLLKGLfexY1l0zOWmpwWbqFZmo4Iks5vAcjsgaJpZM4ZUyxf>
> .
>


-- 
allanc@chickenandporn.com  "金鱼" http://linkedin.com/in/goldfish

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

CharlieDowns commented:
> ...and that happens even if you've seen the text a dozen times; a new set of eyes can find little nit-picky issues. The really nuanced ones are really up to you to accept or refuse -- it's your PR in the end
> -- allanc@chickenandporn.com "金鱼" http://linkedin.com/in/goldfish
Yup and that is why websites like Github is such a great place.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Removed "#Extracting cross compiled packages" part because it turned out it is not needed after some testing.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

chickenandpork commented:
On Jan 6, 2019, at 03:31, CharlieDowns <notifications@github.com> wrote:
> 
> @CharlieDowns commented on this pull request.
> 
> In source/_docs/installation/synology.markdown:
> 
> >  ```bash
> -# mkdir /volume1/homeassistant
> -# chown homeassistant /volume1/homeassistant 
> -# chmod 755 /volume1/homeassistant
> -# cd /volume1/homeassistant
> +# sudo docker run -it -v ~/spksrc:/spksrc synocommunity/spksrc /bin/bash
> Every time the user runs docker run, his container will be removed.
> In case the compilation is cancelled or fails, his docker container would also be removed?
> 
Yes.

My logic in suggesting “—rm” and in keeping a reliable few lines that can be cut-n-pasted is that if the user knew better, he’d choose to drop the —rm.  It causes a user who is less familiar and/or too busy to read in detail a quick way to get the most predictable result.
> Is it not better to let the user decide to remove the containers?
> 
> I might even have to add one anyway in case the user runs docker pull, but the download fails, filling "docker/overlay" with corrupted container
> I tried looking for a command to delete corrupted containers, but "docker rm" did not seem to erase these. I was able to erase these by stopping docker service and deleting "/var/lib/docker", but what are the odds the user is actually using Docker and has to re-download all the Docker containers...?
> 
I’ve never seen this type of failure.  I would expect docker to clean up after itself


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Removed `docker pull` command as docker appears to download the needed container on usage of `docker run`, making that one step less, which is nice.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

CharlieDowns commented:
I have added the old guide, for any users who might want to run Home Assistant with Python 3 provided by Synology, even though they may be stuck on version 0.64.3.
I also updated the warning note at the top of the document for that.

Mr-Meorzipan committed:
Update synology.markdown

nwesterhausen commented:
@CharlieDowns, I saw a typo in a link and a couple other things that I fixed in a pr and some questions about the formatting (copied here):

The code snippets include the leading character '#' or '$' and I assume that was intentional. It breaks just being able to copy paste by triple-clicking, since I have to omit the leading character. And prevents copy/pasting multiple lines at once. However, depending on your view, that can be intended since the "right" way is to just retype it all yourself. 😄

The step-by-step instructions are using un-ordered lists. Are they intentionally not ordered? In my opinion, a step-by-step should use an ordered list since things are done in order.

Thanks.

CharlieDowns commented:
> @CharlieDowns, I saw a typo in a link and a couple other things that I fixed in a pr and some questions about the formatting (copied here):

If possible could you comment on these links? You can do so by highlighting the code and clicking `+`.
In case I know which ones you talk about, the links such as `/docs/installation/docker/#synology-nas` are relative links (e.g., "https://www.home-assistant.io/docs/installation/docker/#synology-nas" or "https://github.com/docs/installation/docker/#synology-nas").

> The code snippets include the leading character '#' or '$' and I assume that was intentional. It breaks just being able to copy paste by triple-clicking, since I have to omit the leading character. And prevents copy/pasting multiple lines at once. However, depending on your view, that can be intended since the "right" way is to just retype it all yourself. smile

It is intentional, when using commands as superuser or root I make sure accidents do not happen by adding a `#` in front of the commands.  
The `$` was added because most of the documentations use it, it makes it easier to spot the commands and looks nice, but the fact you cant simply copy commands is indeed annoying and confusing, will remove it, have not seen it required in the [Standards](https://developers.home-assistant.io/docs/en/documentation_standards.html).

> The step-by-step instructions are using un-ordered lists. Are they intentionally not ordered? In my opinion, a step-by-step should use an ordered list since things are done in order.

That is not intentional, I am using un-ordered list with bullets because it looks nice and most of the documentations seem to use these, but you are right, using a ordered list is more correct.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

nwesterhausen commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,181 @@
-<p class='note warning'>
-Synology only provide Python 3.5.1, which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. You can manually specify the version of Home Assistant to install, for example to install version 0.64.3 you would do `./python3 -m pip install homeassistant==0.64.3`
+<p class="note warning">
+Synology only provides [Python 3.5.1](https://www.synology.com/en-global/dsm/packages/py3k), which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. If you want to run newer Home Assistant versions, you will need to make a new Python 3 package, which will be de ...
```
```suggestion
Synology only provides [Python 3.5.1](https://www.synology.com/en-global/dsm/packages/py3k), which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. If you want to run newer Home Assistant versions, you will need to make a new Python 3 package, which will be detailed in this guide. Otherwise you can follow the instructions "[Using older Python 3 provided by Synology](#using-older-python-3-provided-by-synology)" way below of the document.
```

Mr-Meorzipan replied:
Hmm, yea you are right, when the document is going to be used, the "*linkable_title*" wouldn't exist.

Will correct were used.

nwesterhausen replied:
I think that was the only spot.

Mr-Meorzipan replied:
No, there were other spots, check https://github.com/home-assistant/home-assistant.io/pull/7861/commits/b6cf710e3f08c2bec5a43431f306e334869ca1bb and https://github.com/home-assistant/home-assistant.io/pull/7861/commits/a2e1c1e37ed9aa2f15c5c333bd09855615843ea5

nwesterhausen reviewed (changes requested):
The link just needs to reference the text after "linkable_title" in the title.

nwesterhausen commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,181 @@
-Option 1 is described on the [Docker installation page](/docs/installation/docker/), whereas Option 2 is described below.
+Option 1 is described on the [Docker installation page](/docs/installation/docker/#synology-nas), whereas Option 2 is described below, but as the instructions below require compilation, our recommendation is to run Home Assistant on Docker (If available) or a Raspberri Pi.
 
 
```
```suggestion
```

Mr-Meorzipan replied:
Had already removed a few minutes before this comment, not needed anymore.

Update, nvm, am blind read blank "space", will correct.

nwesterhausen reviewed (commented):
Extra blank line

nwesterhausen commented:
Thanks for the tip about the `+`, I was having trouble figuring out how to add comments on the individual lines. If you haven't already, you should take a look at the preview for the documentation, it gets generated with every commit for this pr:

https://deploy-preview-7861--home-assistant-docs.netlify.com/docs/installation/synology/#using-older-python-3-provided-by-synology

nwesterhausen commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,181 @@
-<p class='note warning'>
-Synology only provide Python 3.5.1, which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. You can manually specify the version of Home Assistant to install, for example to install version 0.64.3 you would do `./python3 -m pip install homeassistant==0.64.3`
+<p class="note warning">
+Synology only provides [Python 3.5.1](https://www.synology.com/en-global/dsm/packages/py3k), which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. If you want to run newer Home Assistant versions, you will need to make a new Python 3 package, which will be de ...
```
```suggestion
Synology only provides [Python 3.5.1](https://www.synology.com/en-global/dsm/packages/py3k), which is not compatible with Home Assistant 0.65.0 or later. Until Synology offer an updated version of Python, Home Assistant 0.64 is the most recent version that will be able to be installed. If you want to run newer Home Assistant versions, you will need to make a new Python 3 package, which will be detailed in this guide. Otherwise you can follow the instructions "[Using older Python 3 provided by Synology](#-linkable_title-using-older-python-3-provided-by-synology-)" at the bottom of this page.
```

Mr-Meorzipan replied:
Already thought about changing that this at work, this consolidates it.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

CharlieDowns commented:
> Thanks for the tip about the `+`, I was having trouble figuring out how to add comments on the individual lines. If you haven't already, you should take a look at the preview for the documentation, it gets generated with every commit for this pr, and you can test the relative links as if it were live.
> 
> https://deploy-preview-7861--home-assistant-docs.netlify.com/docs/installation/synology/#using-older-python-3-provided-by-synology

Oh nice, I did not know about that, thank you.
I will come in handy when trying to make the document look nicer.

CharlieDowns commented:
Btw, is it allowed to use a "<br/>" newline html tag?
The [Standards](https://developers.home-assistant.io/docs/en/documentation_standards.html) says not to make use of ".  " (Markdown inteprets it as newline), but this might also be because it looks quite horrible.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

nwesterhausen commented:
I think it would be encouraged to not insert html except when absolutely necessary. The documentation seems to encourage using long lines. I think what to get what `<br>` would accomplish to just insert a newline into the markdown.

CharlieDowns commented:
> I think it would be encouraged to not insert html except when absolutely necessary. The documentation seems to encourage using long lines. I think what to get what `<br>` would accomplish to just insert a newline into the markdown.

I get what you mean and I use long lines when sentence are a continuation, but sometimes a block become way more readable and nicer to look at if I add it.
I added a few ones, check to see what I mean.

Mr-Meorzipan committed:
Update synology.markdown

Mr-Meorzipan committed:
Update synology.markdown

CharlieDowns commented:
Wow, when you think about it, this pull has gotten quite a lot commits and comments...

Mr-Meorzipan committed:
Update synology.markdown

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
+1. Directly running on DSM
 
-Option 1 is described on the [Docker installation page](/docs/installation/docker/), whereas Option 2 is described below.
+Option 1 is described on the [Docker installation page](/docs/installation/docker/#synology-nas), whereas Option 2 is described below, but as the instructions below require compilation, our recommendation is to run Home Assistant on Docker (If available) or a Raspberri Pi.
```
Home Assistant on Docker (if available) or a Raspberry Pi.

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
 Running these commands will:
+* Compile the required new Python 3 package
+* Install Home Assistant
+* Enable Home Assistant to be launched on http://*Synology_IP*:8123
```
`http://Synology_IP:8123`

DubhAd replied:
`http://Synology_IP:8123`

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
- - Create homeassistant user and add to the "users" group
+Using the [Synology webadmin](https://www.synology.com/en-global/knowledgebase/DSM/help):
+* Install compiled Python 3 package using the [Synology Package Center](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)
+* Create "homeassistant" user
```
Create the `homeassistant` user

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
+Using the [Synology webadmin](https://www.synology.com/en-global/knowledgebase/DSM/help):
+* Install compiled Python 3 package using the [Synology Package Center](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)
+* Create "homeassistant" user
+* Create "homeassistant" [Shared-Folder](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/file_share_desc)
```
Create the `homeassistant` [Shared-Folder](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/file_share_desc)

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,177 @@
-```bash
-# ./python3 -m ensurepip
+Add these edits to make cross compiled Python module packages which Home Assistant needs to install and enable usage of the "[Cloud](/components/cloud/)" and [Homekit](/components/homekit/)" components. Most components will setup just fine, but certain components need a python package that compiles on installation, which will fail as Synology did not provide a compiler, causing said need for cros ...
+Edit "*~/spksrc/spk/python3/src/requirements.txt*" and add the following text to the end of the file:
````
Edit `~/spksrc/spk/python3/src/requirements.txt` and add the following text to the end of the file:

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,177 @@
 
-```bash
-# ./python3 -m ensurepip
+Add these edits to make cross compiled Python module packages which Home Assistant needs to install and enable usage of the "[Cloud](/components/cloud/)" and [Homekit](/components/homekit/)" components. Most components will setup just fine, but certain components need a python package that compiles on installation, which will fail as Synology did not provide a compiler, causing said need for cros ...
````
enable usage of the [Cloud](/components/cloud/) and [Homekit](/components/homekit/) components

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,177 @@
+ed25519==1.4
+```
+
+Add these edits to fix OpenSSL errors when using the "[Xiaomi_Aqara](/components/Xiaomi_Aqara/)" component.<br/>
````
Add these edits to fix OpenSSL errors when using the [Xiaomi_Aqara](/components/Xiaomi_Aqara/) component.<br/>

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,177 @@
+```
+
+Add these edits to fix OpenSSL errors when using the "[Xiaomi_Aqara](/components/Xiaomi_Aqara/)" component.<br/>
+Edit "*~/spksrc/spk/python3/Makefile*" and add the following text above the text "**include ../../mk/spksrc.spk.mk**":
````
Edit `~/spksrc/spk/python3/Makefile` and add the following text above the text `include ../../mk/spksrc.spk.mk`:

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
-Create homeassistant config directory & switch to it
+#### {% linkable_title Compiling the Python 3 package %}
 
+Modify the `docker run` command so "**XXXX**" in "*arch-**XXXX***" contains the appropriate architecture of your Synology NAS. For a list of architectures, look at this [list of architectures](https://github.com/SynoCommunity/spksrc/wiki/Architecture-per-Synology-model) accepted by [spksrc](https://github.com/SynoCommunity/spksrc).<br/>
```
Modify the `docker run` command so `XXXX` in `arch-XXXX` contains the appropriate

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
+#### {% linkable_title Compiling the Python 3 package %}
 
+Modify the `docker run` command so "**XXXX**" in "*arch-**XXXX***" contains the appropriate architecture of your Synology NAS. For a list of architectures, look at this [list of architectures](https://github.com/SynoCommunity/spksrc/wiki/Architecture-per-Synology-model) accepted by [spksrc](https://github.com/SynoCommunity/spksrc).<br/>
+Depending on your computer, compilation may take a hour or more (Significantly less if you have a SSD and a moderately good CPU):
```
hour or more (significantly less

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,177 @@
+# sudo docker run -it -v ~/spksrc:/spksrc -w /spksrc synocommunity/spksrc bash -c 'make clean && make setup && make -C spk/python3 arch-XXXX'
 ```
-Hint: alternatively you can also create a "Shared Folder" via Synology WebUI (e.g., via "File Station") - this has the advantage that the folder is visible via "File Station".
+After the compilation is done, you can find the Python 3 package at "*~/spksrc/packages/python3_XXXX.spk*". It should be named something like "python3_armada370-6.1_3.5.5-7.spk" with a possibly different architecture and version.
````
After the compilation is done, you can find the Python 3 package at `~/spksrc/packages/python3_XXXX.spk`. It should be named something like `python3_armada370-6.1_3.5.5-7.spk` with a possibly different architecture and version

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
+## {% linkable_title Using the Synology webadmin %}
+
+Install the Python 3 package as follows:
+1. Open "[*Package Center*](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)"
```
1. Open [*Package Center*](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
+
+Install the Python 3 package as follows:
+1. Open "[*Package Center*](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)"
+1. Press the "*Manual Install*" button on the top right of the window
```
1. Press the *Manual Install* button on the top right of the window

Same for the the rest of these points down to line 160 - don't put quotes around the option

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -2,78 +2,177 @@
+## {% linkable_title Installing Home Assistant %}
+
+After the Python 3 package has been installed, open terminal and open SSH to the synology.<br/>
+Replace "*user*" with your Synology NAS user and "x.x.x.x" with the its IP address:
```
Replace `user` with your Synology NAS user and `x.x.x.x` with the its IP address:

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -2,78 +2,177 @@
 ```bash
+# sudo /volume1/@appstore/python3/bin/python3 -m pip install homeassistant
+```
+Create a file named "**hass-daemon**" in your "*homeassistant*" Shared-Folder with the script below as its content.<br/>
````
Create a file named `hass-daemon` in your `homeassistant` Shared-Folder with the script below as its content.<br/>

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -175,68 +274,218 @@ case $1 in
+# sudo /volume1/@appstore/python3/bin/python3 -m pip install --upgrade homeassistant
 ```
+<p class="note">
+If you need a update for Python 3 or added a component which failed because of a Python module not installing on your Synology NAS with error code "_Failed building wheel for *Module_Name*_", delete the "*~/spksrc*" directory and redo the guide, but add the failed python module to "*requirements.txt*" from "[Preparing compiling environment](#preparing-compiling-environment)".
````
If you need a update for Python 3 or added a component which failed because of a Python module not installing on your Synology NAS with error code _Failed building wheel for *Module_Name*_, delete the `~/spksrc` directory and redo the guide, but add the failed python module to `requirements.txt` from [Preparing compiling environment](#preparing-compiling-environment).

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
- - Click on OK
- - Click on OK again
+To have Home Assistant start on bootup of your Synology NAS, do as follows:
+1. Open "[*Control Panel*](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/ControlPanel_desc)"
```
Remove quotes here, down to 312

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
+<p class="note warning">
+Synology only provides [Python 3.5.1](https://www.synology.com/en-global/dsm/packages/py3k), which is not compatible with Home Assistant 0.65.0 or later.<br/>
+The instructions above describes how to install newer Home Assistant on your Synology NAS.<br/>
+If you want to proceed with a older version of Home Assistant, follow the instructions below. Please note that you may not be able to use certain components, such as "[Cloud](/components/cloud/)" and [Homekit](/components/homekit/)".
```
Remove quotes

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
-- Start Home Assistant:
+Running these commands will:
+* Install Home Assistant 0.64.3
+* Enable Home Assistant to be launched on http://*Synology_IP*:8123
```
`http://Synology_IP:8123`

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -175,68 +274,218 @@ case $1 in
-```
+Using the [Synology webadmin](https://www.synology.com/en-global/knowledgebase/DSM/help):
+* Install Python 3 using the Synology Package Center
+* Create "homeassistant" user
````
* Create the `homeassistant` user

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
+Using the [Synology webadmin](https://www.synology.com/en-global/knowledgebase/DSM/help):
+* Install Python 3 using the Synology Package Center
+* Create "homeassistant" user
+* Create "homeassistant" [Shared-Folder](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/file_share_desc)
```
Create the `homeassistant` [Shared-Folder]

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
+### {% linkable_title Install Synology provided Python 3 %}
+
+Install Python 3 as follows:
+1. Open "[*Package Center*](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/PkgManApp/PackageCenter_desc)"
```
Remove quotes here and next two lines

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
+### {% linkable_title Installing Home Assistant 0.64.3 %}
 
+After the Python 3 package has been installed, open terminal and open SSH to the synology.<br/>
+Replace "*user*" with your Synology NAS user and "x.x.x.x" with the its IP address:
```
Replace `user` with your Synology NAS user and `x.x.x.x` with the its IP address:

DubhAd commented on source/_docs/installation/synology.markdown:
````diff
@@ -175,68 +274,218 @@ case $1 in
+```bash
+# sudo /volume1/@appstore/py3k/usr/local/bin/python3 -m pip install homeassistant==0.64.3
+```
+Create a file named "**hass-daemon**" in your "*homeassistant*" Shared-Folder with the script below as its content.
````
Create a file named `hass-daemon` in your `homeassistant` Shared-Folder with the script below as its content.

DubhAd commented on source/_docs/installation/synology.markdown:
```diff
@@ -175,68 +274,218 @@ case $1 in
+### {% linkable_title Starting Home Assistant 0.64.3 on bootup %}
+
+To have Home Assistant start on bootup of your Synology NAS, do as follows:
+1. Open "[*Control Panel*](https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/ControlPanel_desc)"
```
Remove quotes here to the end

frenck commented:
Since there was no response in almost 2 months, I'm declaring this PR stale and closing it.
Feel free to re-open this PR when you feel like that is the appropriate thing to do.
